compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3108,12 +3115,17 @@ class Typer extends Namer
         val propFail = propagatedFailure(args)
 
         def issueErrors(): Tree = {
+          lazy val paramSyms = tree.symbol.paramSymss.flatten.map(sym => sym.name -> sym).toMap
+
           wtp.paramNames.lazyZip(wtp.paramInfos).lazyZip(args).foreach { (paramName, formal, arg) =>
             arg.tpe match {
               case failure: SearchFailureType =>
+                val paramSymWithMethodTree = if tree.symbol.exists then Some((paramSyms(paramName), tree)) else None
+
                 report.error(
```

Comment:
In many situation we will recover from an error without ever showing it to the user (e.g., in implicit search we might find another candidate), so it's important to delay computing anything we want to show to the user until we know we need it. Here things are a bit subtle but `missingArgMsg` returns a String and `report.error` takes as input a `Message`, this works because there is an implicit conversion from String to Message: https://github.com/lampepfl/dotty/blob/22c23a5cc648650bb6380dec0c11a5fdc34e1d73/compiler/src/dotty/tools/dotc/reporting/Message.scala#L15
Note that this conversion takes a `=> String` as input and which means the call to `missingArgMsg` will be delayed, which is what we want. But it would be even better if `paramSyms` and `paramSymWithMethodTree` were also delayed, which can be achieved by computing them inside `missingArgMsg` (or making them both lazy vals)